### DISCUSSION TITLE
Fatfs partition not recognized in ESP32

### DISCUSSION THREAD

**Discussion by danimasa** [micropython/micropython#9152] [Using MicroPython] [answered]
Hello

I'm trying to mount a FATFS filesystem in an ESP32, I generate the binary and upload it to the vfs partition but if I try to do a simple mp_vfs_stat I keep receiving `NLR jump failed, val=0x3ffe58b0` error.

Isn't the vfs partition mounted in root fs at startup?
Frozen files are kept in a different partition? Because `boot.py` is running correctly

**Comment by jimmo** [comment, marked as answer]
> I'm trying to mount a FATFS filesystem in an ESP32, I generate the binary and upload it to the vfs partition but if I try to do a simple mp_vfs_stat I keep receiving NLR jump failed, val=0x3ffe58b0 error.

The NLR jump failed shouldn't be possible no matter what you're doing, so it sounds like you're definitely hitting a bug somewhere if you're only using Python.

I'm a little bit unsure what you mean though. What does "generate the binary and upload it to the vfs partition". Are you generating a FAT filesystem and then using esptool to write it to the correct offset in flash?

`mp_vfs_stat` would suggest that you're writing something in C. So I think we need more info -- where are you calling this from?

> Isn't the vfs partition mounted in root fs at startup?

Yes. See https://github.com/micropython/micropython/blob/master/ports/esp32/modules/_boot.py -- the flashbdev.py uses `Partition.find(Partition.TYPE_DATA, label="vfs")`.

> Frozen files are kept in a different partition? Because boot.py is running correctly

Frozen files are not in an ESP32 partition. Rather they're part of the application code (the freezing process turns it into a .c file which is compiled and linked).

**Reply by jimmo** [reply]
Another thought, `NLR jump failed` would likely be caused by calling into MicroPython code (that's raising an exception) from a context that isn't set up for NLR (i.e. MicroPython's exception handling). i.e. are you in a different FreeRTOS task maybe?

**Reply by danimasa** [reply]
> The NLR jump failed shouldn't be possible no matter what you're doing, so it sounds like you're definitely hitting a bug somewhere if you're only using Python.

I said that because I traced the error and it's ocurring when I try to read some data from the filesystem. One of the cases was trying to execute the `mp_vfs_stat` function.

> I'm a little bit unsure what you mean though. What does "generate the binary and upload it to the vfs partition". Are you generating a FAT filesystem and then using esptool to write it to the correct offset in flash?

Yes, it's just that. I used the esp-idf 5.0 `fatfsgen` tool to generate de image and uploaded with esptool to the correct offset partition.

> mp_vfs_stat would suggest that you're writing something in C. So I think we need more info -- where are you calling this from?

Yes I'm trying to adapt the micropython to run python files from a C written application. Basically I'll call some functions from python files based on some logic. I managed to read files from a SPIFFS filesystem, but as micropython only implements native adaptation to FAT filesystems I'm trying to port without success. I'll need to execute python files and read yaml files too.

**Reply by danimasa** [reply]
> Another thought, `NLR jump failed` would likely be caused by calling into MicroPython code (that's raising an exception) from a context that isn't set up for NLR (i.e. MicroPython's exception handling). i.e. are you in a different FreeRTOS task maybe?

I think not, my code is like this:

```cpp
...

  volatile uint32_t sp = (uint32_t)get_sp();
  mp_thread_init(pxTaskGetStackStart(NULL), MP_TASK_STACK_SIZE / sizeof(uintptr_t));
  mp_stack_set_top((void *)sp);
  mp_stack_set_limit(MP_TASK_STACK_SIZE - MP_TASK_STACK_LIMIT_MARGIN);
  gc_init(mp_task_heap, mp_task_heap + mp_task_heap_size);
  mp_init();

  pyexec_frozen_module("_boot.py");
  pyexec_file_if_exists("boot.py");

  const char* path = "/ports.yml";
  mp_obj_t fl_stat = mp_vfs_stat(mp_obj_new_str(path, strlen(path)));
  printf("    %s", mp_obj_str_get_str(fl_stat));
```

I'm not creating a new task

**Reply by jimmo** [reply]
> I'm not creating a new task

Ah yes, this is exactly what I meant -- you're at the top-level of the MicroPython task, but not inside an exception handling context.

Here's how to write `try/catch` in MicroPython C.  (see for example the parse_compile_execute function which runs a Python file).

```c
nlr_buf_t nlr;
nlr.ret_val = NULL;
if (nlr_push(&nlr) == 0) {
   // Try
   ... mp_vfs_stat... etc
} else {
   // Except
   mp_obj_print_exception(&mp_plat_print, MP_OBJ_FROM_PTR(nlr.ret_val));
}
```

Whatever exception is being raised might give you some hints as to what's going wrong.

**Reply by danimasa** [reply]
Really, after putting the code in a exception handling context a different error appears: `OSError: [Errno 2] ENOENT` but I yet don't understand the reason of the error.

**Reply by robert-hh** [reply]
ENOENT is raised, when a file is not found. Did you create the file system on the partition?

**Reply by danimasa** [reply]
> ENOENT is raised, when a file is not found. Did you create the file system on the partition?

I was expecting that the boot.py would create it to me, because of it I'm running it first. Should I have to create it using the esp32 libraries?

Really the file was not there, I used the `mp_vfs_listdir` to list the files in the root and it only shows the `boot.py` file.

**Reply by karfas** [reply]
Please **read** the (frozen) modules in https://github.com/micropython/micropython/tree/master/ports/esp32/modules.

The default _boot.py searches (in flashbdev.py) for a partition labeled either "vfs" or "ffat" (from esp32/partitions*.csv) and happily creates a LFS2 file system here (even for "ffat" labeled partitions) if no filesystem already exists.
If micropython should create a FAT fs for you, you will need to alter inisetup.py accordingly.
If you generate the FAT fs as an image, you need to flash your filesystem to a correct labeled partition.

**Reply by danimasa** [reply]
I found the problem, the image I was uploading to esp32 was corrupted, I generated the image again and now the _boot.py was capable to mount it correctly. Thank you.
